### PR TITLE
Define error

### PR DESCRIPTION
**Pull request by cjfghk5697**
#2069 

I have implemented the suggestion from issue #2069. The `FileMetadataError` class has been moved from `huggingface_hub/inference/_templating.py` to `huggingface_hub/errors.py`.

Could you please review this PR? If this is correct, I will move the other classes in the same way.

### CODE REVIEW

**Comment by cjfghk5697** [comment]
@Wauplin Are there any other error classes that need to be modified or moved besides `FileMetadataError`?  Also, do you think there's a need for any error handling code for 500-level errors?

**Comment by 010kim** [comment]
Hello, I was going over the same code actually. It looks like FileMetadataError is the only error that is not affected by hf_raise_for_status, so all the other errors _errors.py should remain there. @Wauplin So I won't create any separate PR. Thanks a lot for your response!

File: src/huggingface_hub/file_download.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 from dataclasses import dataclass
 from pathlib import Path
 from typing import Any, BinaryIO, Dict, Literal, NoReturn, Optional, Tuple, Union
+from huggingface_hub.errors import FileMetadataError
```
Comment (Wauplin) [inline src/huggingface_hub/file_download.py:17]:
```suggestion
from .errors import FileMetadataError
```

**Review by Wauplin — COMMENTED** [review]
Hi @cjfghk5697 , thanks for your PR! Actually all errors from `src/huggingface_hub/utils/_errors.py` are entitled to move to `src/huggingface_hub/errors.py` as mentioned in https://github.com/huggingface/huggingface_hub/issues/2069#issuecomment-2284038080. The logic for `hf_raise_for_status` stays where it is but the error classes definitions have to move. It can be done in this PR or a separate one. 

> Also, do you think there's a need for any error handling code for 500-level errors?

I don't think there is a special need for that no. Those are usually caught as `HfHttpError` and then filtered using the status code directly.

**Comment by Wauplin** [comment]
> Hello, I was going over the same code actually. It looks like FileMetadataError is the only error that is not affected by hf_raise_for_status, so all the other errors _errors.py should remain there. @Wauplin So I won't create any separate PR. Thanks a lot for your response!

@010kim sorry for the duplicate work. To avoid such a situation, would you like to focus on moving the 2 exceptions from [src/huggingface_hub/utils/_cache_manager.py](https://github.com/huggingface/huggingface_hub/blob/a99071d866893e36c2799be446470d1e650cc9d5/src/huggingface_hub/utils/_cache_manager.py#L37) for example? They also have to be moved in a separate PR similar to this one.

**Comment by 010kim** [comment]
Of course! That's great, I'm on it!

**Comment by 010kim** [comment]
> > Hello, I was going over the same code actually. It looks like FileMetadataError is the only error that is not affected by hf_raise_for_status, so all the other errors _errors.py should remain there. @Wauplin So I won't create any separate PR. Thanks a lot for your response!
> 
> @010kim sorry for the duplicate work. To avoid such a situation, would you like to focus on moving the 2 exceptions from [src/huggingface_hub/utils/_cache_manager.py](https://github.com/huggingface/huggingface_hub/blob/a99071d866893e36c2799be446470d1e650cc9d5/src/huggingface_hub/utils/_cache_manager.py#L37) for example? They also have to be moved in a separate PR similar to this one.

Of course! That's great, I'm on it!

**Comment by cjfghk5697** [comment]
@Wauplin, thank you for your reply! Do you have any recommended suggestions or issues for improving Hub?

**Comment by Wauplin** [comment]
Hi @cjfghk5697, thanks for asking! Sorry for the noise on your PR, let's refocus on this one first :wink: As mentioned in https://github.com/huggingface/huggingface_hub/pull/2444#pullrequestreview-2236019628, it would be great to move all errors from `src/huggingface_hub/utils/_errors.py` to `src/huggingface_hub/errors.py`, not only `FileMetadataError`. The `hf_raise_for_status` should stay where it is. Could you have a look at that? :pray: :hugs:

**Comment by cjfghk5697** [comment]
@Wauplin Thank you! I'm a bit confused should the class used in hf_raise_for_status also be moved to error.py? For example, the RevisionNotFoundError class is used in hf_raise_for_status, but should it also be moved to error.py?

**Comment by Wauplin** [comment]
>  I'm a bit confused should the class used in hf_raise_for_status also be moved to error.py? For example, the RevisionNotFoundError class is used in hf_raise_for_status, but should it also be moved to error.py?🤗

Yes, all of those classes! (`RevisionNotFoundError`, `GatedRepoError`, etc.)
Then, you'll have to import them back from `src/huggingface_hub/errors.py` in `src/huggingface_hub/utils/_errors.py`. The goal is to have all classes defined in the same `errors` module but we should still keep some logic outside (where it is currently)

**Comment by cjfghk5697** [comment]
@Wauplin I got it and appreciate your reply😊 The classes in `_errors.py` have been moved to `errors.py`. Could you check the file?  🤗.

File: src/huggingface_hub/utils/_errors.py [hunk 2]
```diff
@@ -1,250 +1,19 @@
...
-        ^https://[^/]+
-        (
-            # on /api/repo_type/repo_id
-            /api/(models|datasets|spaces)/(.+)
-            |
-            # or /repo_id/resolve/revision/...
-            /(.+)/resolve/(.+)
-        )
-    """,
-    flags=re.VERBOSE,
```
Comment (Wauplin) [inline src/huggingface_hub/utils/_errors.py:21]: This can be kept in `./utils/_errors.py` since it's not a Exception class definition

**Review by Wauplin — COMMENTED** [review]
Thanks @cjfghk5697 for the changes! Almost everything looks good to me for a merge. Left a minor comment and then we should be good to go! :rocket:

**Comment by cjfghk5697** [comment]
Thanks @Wauplin! I have put it back!

**Review by Wauplin — APPROVED** [review]
Perfect!

**Comment by Wauplin** [comment]
Failing tests are unrelated to this PR so let's merge it :)

**Comment by cjfghk5697** [comment]
@Wauplin 
Hello! Is there anything else I can assist you with besides handling this issue? I'd be happy to offer additional support alongside the HuggingFace OSSCA team!

**Comment by Wauplin** [comment]
Hi @cjfghk5697, thanks again for your support! What do you say about tackling https://github.com/huggingface/huggingface_hub/issues/2466? :) If not, you can always have a look to the list in https://github.com/huggingface/huggingface_hub/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22.

**Comment by cjfghk5697** [comment]
Hi @Wauplin, thanks for the suggestion! I’ll go ahead and tackle #2466. Looking forward to working on it!

**Pull request merged**